doc/subsystems/nvs/nvs.rst

```diff
@@ -0,0 +1,95 @@
+.. _nvs:
+
+Non-volatile storage (NVS) for zephyr
+#####################################
+
+Elements, represented as id-data pairs, are stored in flash using a
+FIFO-managed circular buffer. The flash area is divided into sectors. Elements
+are appended to a sector until storage space in the sector is exhausted. Then a
+new sector in the flash area is prepared for use (erased). Before erasing the
+sector it is checked that identifier - data pairs exist in the sectors in use,
+if not the id-data pair is copied.
+
+The id is a 16-bit unsigned number, where two values are reserved:
+
+- ``0xFFFF`` is used to determine free locations in flash
+- ``0xFFFE`` is used to close a sector
+
+NVS ensures that for each id there is at least one id-data pair stored in flash
+at all time.
+
+NVS does flash wear leveling, a technique for prolonging the service life of
+flash memory components.  Each new write is done at the next available flash
+location.  Subsequent requested writes of the same id-data value are not
+actually written to the flash.  When using the high-level APIs to save all
+values, only the ones that would change are actually written to flash.
+
+NVS allows storage of binary blobs, strings, integers, longs, and any
+combination of these. As each element is stored with a header (containing the
+id and length) and a footer (containing the crc16) it is less suited to store
+only integers.
+
+For NVS the file system is declared as:
+
+.. code-block:: c
+static struct nvs_fs fs = {
+  .sector_size = SECTOR_SIZE,
+  .sector_count = SECTOR_COUNT,
+  .offset = STORAGE_OFFSET,
+};
+
+where
+
+.. code-block:: c
+  SECTOR_SIZE is a multiple of the flash erase page size and a power of 2,
+  SECTOR_COUNT is at least 2, one sector is always kept empty to allow copying
+  of existing data.
+  STORAGE_OFFSET is the offset of the storage area in flash.
+
+Flash wear
+**********
+
+When writing data to flash a study of the flash wear is important. Flash has a
+limited life which is determined by the number of times flash can be erased.
+Flash is erased one page at a time and the pagesize is determined by the
+hardware. As an example a nRF51822 device has a pagesize of 1024 bytes and each
+page can be erased about 20000 times.
+
+So how to calculate expected device life when using NVS:
```

Comment:
I think would would read better just as a title for a sub-section, so change to:

    Calculating expected device lifetime
    ====================================